src/Http/Routing/src/RouteEndpointDataSource.cs

```diff
@@ -225,6 +229,22 @@ internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
             builder.RequestDelegate = factoryCreatedRequestDelegate;
         }
 
+        entry.FinallyConventions.IsReadonly = true;
+        for (var i = entry.FinallyConventions.Count - 1; i >= 0; i--)
+        {
+            entry.FinallyConventions[i](builder);
+        }
+
+        if (groupFinallyConventions is not null)
+        {
+            // Group conventions are ordered by the RouteGroupBuilder before
+            // being provided here.
+            foreach (var groupFinallyConvention in groupFinallyConventions)
```

Comment:
We got it right here at least. I'm now wondering about the wisdom of pre-reversing this list. It'd make more sense if the entry-specific finally conventions were stored in the stack instead of a List.

I no longer care where we reverse these. But we need regression tests that make sure we always run group and endpoint-specific `Finally` callbacks for all the `EndpointDataSource`'s we've modifed.

Reply:
> but we do want the LIFO ordering. But we need regression tests that make sure we always run group and endpoint-specific Finally callbacks in LIFO order for all the EndpointDataSource's we've modifed.

`FinallyONestednGroups_CanExamineFinallyInLIFOOrder` does this. I've updated it to include finally conventions on nested groups _and_ multiple finally conventions on an endpoint.